R/plumber.R

```diff
@@ -890,29 +802,30 @@ plumber <- R6Class(
       filter <- PlumberFilter$new(name, expr, private$envir, serializer)
       private$addFilterInternal(filter)
     },
-    #' @description Retrieve openAPI file
-    swaggerFile = function() { #FIXME: test
-
-      swaggerPaths <- private$swaggerFileWalkMountsInternal(self)
-
-      # Extend the previously parsed settings with the endpoints
-      def <- utils::modifyList(private$globalSettings, list(paths = swaggerPaths))
-
-      # Lay those over the default globals so we ensure that the required fields
-      # (like API version) are satisfied.
+    #' @description Retrieve OpenAPI specification
+    openAPISpec = function() {
+      routerSpec <- private$routerSpecificationInternal(self)
+      def <- utils::modifyList(private$globalSettings, list(paths = routerSpec))
       ret <- utils::modifyList(defaultGlobals, def)
-
-      # remove NA or NULL values, which swagger doesn't like
+      if (is.list(self$customSpec)) {
+        ret <- utils::modifyList(ret, self$customSpec)
+      }
       ret <- removeNaOrNulls(ret)
-
       ret
     },
-    #' @description Retrieve openAPI file
-    openAPIFile = function() {
-      self$swaggerFile()
-    },
+    #' @field customSpec A list of custom spec to overlay over openAPI spec
+    #' generated from a plumber router.
+    customSpec = NULL,
```

Comment:
Rather than using a static object, I'd like to supply a function directly.

```r
#### in the private defs
    apiHandler = force,
```

```r
    #' @description Add a function to customize what is returned in `$apiSpec()`. 
    #' @param api_fun This function should accept an OpenAPI Specification autogenerated by `plumber`.  It should return a list object in valid OpenAPI Specification format. This will not be validated.
    setApiHandler = function(api_fun = force) {
      # allow null values
      if (!is.null(api_fun)) {
        stopifnot(is.function(api_fun))
      }
      private$apiHandler <- api_fun
    }
```